### PR TITLE
UAR-975 / UAR-1357 Enable leading zero and days and months in dates

### PR DESCRIPTION
**Pull request by AndreKramer7654**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-975 
https://companieshouse.atlassian.net/browse/UAR-1357

### Change description

Date validation was tightened because of several service unavailability crashes due to day or month input.
This PR relaxes the date validation again to allow leading zeros on days and months.
Will need comprehensive testing to verify that no 01/01/XYZW dates reintroduce service unavailable exceptions.

copilot:all

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/validation/custom.validation.ts [hunk 1]
```diff
@@ -68,10 +68,14 @@ export const checkDateIsWithinLast3Months = (errMsg: string, day: string = "", m
...
 };
 
+const digitsWithLeadingZero = ["01", "02", "03", "04", "05", "06", "07", "08", "09"];
+
 export const checkDateValueIsValid = (invalidDateErrMsg: string, dayStr: string = "", monthStr: string = "", yearStr: string = "") => {
   const day = parseInt(dayStr), month = parseInt(monthStr), year = parseInt(yearStr);
   if (isNaN(day) || isNaN(month) || isNaN(year) ||
-    day.toString() !== dayStr || month.toString() !== monthStr || year.toString() !== yearStr ||
+    (day.toString() !== dayStr && !digitsWithLeadingZero.includes(dayStr)) ||
+    (month.toString() !== monthStr && !digitsWithLeadingZero.includes(monthStr)) ||
```
Comment (ConnorOS-ch) [inline src/validation/custom.validation.ts:77]: This change works, although I think the use of regex would be simpler

```const day = parseInt(dayStr), month = parseInt(monthStr), year = parseInt(yearStr);
  const dayMonthRegex = /^\d{1,2}$/;
  const yearRegex = /^\d{4}$/;

  if (isNaN(day) || isNaN(month) || isNaN(year) ||
    !dayMonthRegex.test(dayStr) || !dayMonthRegex.test(monthStr) || !yearRegex.test(yearStr) ||
    !DateTime.utc(year, month, day).isValid) {
    throw new Error(invalidDateErrMsg);
  }
```

Comment (markpit) [reply src/validation/custom.validation.ts:80]: used the regex for day/month but year 0999 was failing in unit test so have used the original year check.
This is also inline with our new story to allow leading zero in day and month - https://companieshouse.atlassian.net/browse/UAR-1357

Comment (markpit) [reply src/validation/custom.validation.ts:80]: push didn't complete, will try again...

**Review by mattch1 — APPROVED** [review]
LGTM - might be worth adding some comments to the "test each" tests just emphasising that they are testing valid and invalid date formats.

File: test/validation/date.validation.spec.ts [hunk 2]
```diff
@@ -587,29 +600,4 @@ describe("should chek date functions for custom validation", () => {
   test("should test checkIdentityDate returns true for no param", () => {
     expect(checkIdentityDate()).toBe(true);
   });
-
-  test("should test 1/1/2001 returns true", () => {
```
Comment (mwestacott) [inline test/validation/date.validation.spec.ts:591]: Should this still be returning true or an error? If true, is it covered by another test somewhere?

Comment (markpit) [reply test/validation/date.validation.spec.ts:591]: i can add it in with the other true tests

Comment (markpit) [reply test/validation/date.validation.spec.ts:591]: updated

Comment (mwestacott) [reply test/validation/date.validation.spec.ts:591]: Thanks.

**Review by mwestacott — APPROVED** [review]

**Review by dlloyd1-ch — APPROVED** [review]

**Pull request merged**